// @ionic-native/in-app-purchase/ngx/index.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { IonicNativePlugin, cordova } from '@ionic-native/core';
var InAppPurchase = /** @class */ (function (_super) {
    __extends(InAppPurchase, _super);
    function InAppPurchase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    InAppPurchase.prototype.getProducts = function (productId) { return cordova(this, "getProducts", { "otherPromise": true }, arguments); };
    InAppPurchase.prototype.buy = function (productId) { return cordova(this, "buy", { "otherPromise": true }, arguments); };
    InAppPurchase.prototype.subscribe = function (productId) { return cordova(this, "subscribe", { "otherPromise": true }, arguments); };
    InAppPurchase.prototype.consume = function (productType, receipt, signature) { return cordova(this, "consume", { "otherPromise": true }, arguments); };
    InAppPurchase.prototype.restorePurchases = function () { return cordova(this, "restorePurchases", { "otherPromise": true }, arguments); };
    InAppPurchase.prototype.getReceipt = function () { return cordova(this, "getReceipt", { "otherPromise": true, "platforms": ["iOS"] }, arguments); };
    InAppPurchase.pluginName = "InAppPurchase";
    InAppPurchase.plugin = "cordova-plugin-inapppurchase";
    InAppPurchase.pluginRef = "inAppPurchase";
    InAppPurchase.repo = "https://github.com/AlexDisler/cordova-plugin-inapppurchase";
    InAppPurchase.platforms = ["Android", "iOS"];
    InAppPurchase = __decorate([
        Injectable()
    ], InAppPurchase);
    return InAppPurchase;
}(IonicNativePlugin));
export { InAppPurchase };
