// @ionic-native/in-app-purchase/index.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { IonicNativePlugin, cordova } from '@ionic-native/core';
var InAppPurchaseOriginal = /** @class */ (function (_super) {
    __extends(InAppPurchaseOriginal, _super);
    function InAppPurchaseOriginal() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    InAppPurchaseOriginal.prototype.getProducts = function (productId) { return cordova(this, "getProducts", { "otherPromise": true }, arguments); };
    InAppPurchaseOriginal.prototype.buy = function (productId) { return cordova(this, "buy", { "otherPromise": true }, arguments); };
    InAppPurchaseOriginal.prototype.subscribe = function (productId) { return cordova(this, "subscribe", { "otherPromise": true }, arguments); };
    InAppPurchaseOriginal.prototype.consume = function (productType, receipt, signature) { return cordova(this, "consume", { "otherPromise": true }, arguments); };
    InAppPurchaseOriginal.prototype.restorePurchases = function () { return cordova(this, "restorePurchases", { "otherPromise": true }, arguments); };
    InAppPurchaseOriginal.prototype.getReceipt = function () { return cordova(this, "getReceipt", { "otherPromise": true, "platforms": ["iOS"] }, arguments); };
    InAppPurchaseOriginal.pluginName = "InAppPurchase";
    InAppPurchaseOriginal.plugin = "cordova-plugin-inapppurchase";
    InAppPurchaseOriginal.pluginRef = "inAppPurchase";
    InAppPurchaseOriginal.repo = "https://github.com/AlexDisler/cordova-plugin-inapppurchase";
    InAppPurchaseOriginal.platforms = ["Android", "iOS"];
    return InAppPurchaseOriginal;
}(IonicNativePlugin));
var InAppPurchase = new InAppPurchaseOriginal();
export { InAppPurchase };
